File: php/phpsearch/searchoptions.php
<?php

require_once __DIR__ . '/autoload.php';

/**
 * Class SearchOptions
 */
class SearchOptions
{
    /**
     * @var array
     */
    private $options;

    public function __construct()
    {
        $this->options = array();

        $this->arg_action_map = [
            'encoding' => function (string $s, SearchSettings $settings) {
                $settings->textfileencoding = $s;
            },
            'in-archiveext' => function (string $s, SearchSettings $settings) {
                $settings->add_exts($s, $settings->in_archiveextensions);
            },
            'in-archivefilepattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->in_archivefilepatterns);
            },
            'in-dirpattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->in_dirpatterns);
            },
            'in-ext' => function (string $s, SearchSettings $settings) {
                $settings->add_exts($s, $settings->in_extensions);
            },
            'in-filepattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->in_filepatterns);
            },
            'in-filetype' => function (string $s, SearchSettings $settings) {
                $settings->add_filetypes($s, $settings->in_filetypes);
            },
            'in-linesafterpattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->in_linesafterpatterns);
            },
            'in-linesbeforepattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->in_linesbeforepatterns);
            },
            'linesafter' => function (string $s, SearchSettings $settings) {
                $settings->linesafter = intval($s);
            },
            'linesaftertopattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->linesaftertopatterns);
            },
            'linesafteruntilpattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->linesafteruntilpatterns);
            },
            'linesbefore' => function (string $s, SearchSettings $settings) {
                $settings->linesbefore = intval($s);
            },
            'maxlinelength' => function (string $s, SearchSettings $settings) {
                $settings->maxlinelength = intval($s);
            },
            'out-archiveext' => function (string $s, SearchSettings $settings) {
                $settings->add_exts($s, $settings->out_archiveextensions);
            },
            'out-archivefilepattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->out_archivefilepatterns);
            },
            'out-dirpattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->out_dirpatterns);
            },
            'out-ext' => function (string $s, SearchSettings $settings) {
                $settings->add_exts($s, $settings->out_extensions);
            },
            'out-filepattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->out_filepatterns);
            },
            'out-filetype' => function (string $s, SearchSettings $settings) {
                $settings->add_filetype($s, $settings->out_filetypes);
            },
            'out-linesafterpattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->out_linesafterpatterns);
            },
            'out-linesbeforepattern' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->out_linesbeforepatterns);
            },
            'search' => function (string $s, SearchSettings $settings) {
                $settings->add_patterns($s, $settings->searchpatterns);
            },
            'settings-file' => function (string $s, SearchSettings $settings) {
                $this->settings_from_file($s, $settings);
            }
        ];

        $this->bool_flag_action_map = [
            'allmatches' => function (bool $b, SearchSettings $settings) {
                $settings->firstmatch = !$b;
            },
            'archivesonly' => function (bool $b, SearchSettings $settings) {
                $settings->set_archivesonly($b);
            },
            'debug' => function (bool $b, SearchSettings $settings) {
                $settings->set_debug($b);
            },
            'excludehidden' => function (bool $b, SearchSettings $settings) {
                $settings->excludehidden = $b;
            },
            'firstmatch' => function (bool $b, SearchSettings $settings) {
                $settings->firstmatch = $b;
            },
            'help' => function (bool $b, SearchSettings $settings) {
                $settings->printusage = $b;
            },
            'includehidden' => function (bool $b, SearchSettings $settings) {
                $settings->excludehidden = !$b;
            },
            'listdirs' => function (bool $b, SearchSettings $settings) {
                $settings->listdirs = $b;
            },
            'listfiles' => function (bool $b, SearchSettings $settings) {
                $settings->listfiles = $b;
            },
            'listlines' => function (bool $b, SearchSettings $settings) {
                $settings->listlines = $b;
            },
            'multilinesearch' => function (bool $b, SearchSettings $settings) {
                $settings->multilinesearch = $b;
            },
            'noprintmatches' => function (bool $b, SearchSettings $settings) {
                $settings->printresults = !$b;
            },
            'norecursive' => function (bool $b, SearchSettings $settings) {
                $settings->recursive = !$b;
            },
            'nosearcharchives' => function (bool $b, SearchSettings $settings) {
                $settings->searcharchives = !$b;
            },
            'printmatches' => function (bool $b, SearchSettings $settings) {
                $settings->printresults = $b;
            },
            'recursive' => function (bool $b, SearchSettings $settings) {
                $settings->recursive = $b;
            },
            'searcharchives' => function (bool $b, SearchSettings $settings) {
                $settings->searcharchives = $b;
            },
            'uniquelines' => function (bool $b, SearchSettings $settings) {
                $settings->uniquelines = $b;
            },
            'verbose' => function (bool $b, SearchSettings $settings) {
                $settings->verbose = $b;
            },
            'version' => function (bool $b, SearchSettings $settings) {
                $settings->printversion = $b;
            }
        ];
        $this->longarg_map = array();
        $this->set_options_from_xml();
    }

    private function set_options_from_xml()
    {
        $searchoptionspath = FileUtil::expand_user_home_path(Config::SEARCHOPTIONSPATH);
        if (file_exists($searchoptionspath)) {
            $searchoptions = simplexml_load_file($searchoptionspath);
            foreach ($searchoptions->searchoption as $searchoption) {
                $short = sprintf($searchoption['short']);
                $long = sprintf($searchoption['long']);
                $desc = trim($searchoption);
                $func = null;
                if (array_key_exists($long, $this->arg_action_map)) {
                    $func = $this->arg_action_map[$long];
                } elseif (array_key_exists($long, $this->bool_flag_action_map)) {
                    $func = $this->bool_flag_action_map[$long];
                }
                $option = new SearchOption($short, $long, $desc, $func);
                $this->options[] = $option;
                $this->longarg_map[$long] = $long;
                if ($short) {
                    $this->longarg_map[$short] = $long;
                }
            }
            usort($this->options, 'cmp_searchoptions');
        } else {
            throw new Exception('File not found: ' . $searchoptionspath);
        }
    }

    private function settings_from_file(string $filepath, SearchSettings $settings)
    {
        if (!file_exists($filepath)) {
            throw new SearchException('Settings file not found');
        }
        $json = file_get_contents($filepath);
        $this->settings_from_json($json, $settings);
    }

    public function settings_from_json(string $json, SearchSettings $settings)
    {
        $json_obj = json_decode($json, true);
        foreach (array_keys($json_obj) as $k) {
            if (array_key_exists($k, $this->arg_action_map)) {
                $this->arg_action_map[$k]($json_obj[$k], $settings);
            } elseif (array_key_exists($k, $this->bool_flag_action_map)) {
                $this->bool_flag_action_map[$k]($json_obj[$k], $settings);
            } elseif ($k == 'startpath') {
                $settings->startpath = $json_obj[$k];
            } else {
                throw new SearchException("Invalid option: $k");
            }
        }
    }

    public function settings_from_args(array $args): SearchSettings
    {
        $settings = new SearchSettings();
        while (count($args) > 0) {
            $arg = array_shift($args);
            if ($arg{0} == '-') {
                while ($arg{0} == '-') {
                    $arg = substr($arg, 1);
                }
                $longarg = $this->longarg_map[$arg];
                if (array_key_exists($longarg, $this->arg_action_map)) {
                    if (count($args) > 0) {
                        $val = array_shift($args);
                        $this->arg_action_map[$longarg]($val, $settings);
                    } else {
                        throw new SearchException("Missing value for $arg");
                    }
                } elseif (array_key_exists($longarg, $this->bool_flag_action_map)) {
                    $this->bool_flag_action_map[$longarg](true, $settings);
                    if (in_array($longarg, array("help", "version"))) {
                        break;
                    }
                } else {
                    throw new SearchException("Invalid option: $arg");
                }
            } else {
                $settings->startpath = $arg;
            }
        }
        return $settings;
    }

    public function usage()
    {
        echo $this->get_usage_string() . "\n";
    }

    private function get_usage_string(): string
    {
        $usage = "Usage:\n phpsearch.php [options] -s <searchpattern>";
        $usage .= " <startpath>\n\nOptions:\n";
        $opt_map = array();
        $longest = 0;
        foreach ($this->options as $option) {
            $opt_str = '';
            if ($option->shortarg) {
                $opt_str = '-' . $option->shortarg . ',';
            }
            $opt_str .= '--' . $option->longarg;
            if (strlen($opt_str) > $longest) {
                $longest = strlen($opt_str);
            }
            $opt_map[$opt_str] = $option->desc;
        }
        $format_str = " %-" . $longest . "s  %s\n";
        foreach ($opt_map as $opt => $desc) {
            $usage .= sprintf($format_str, $opt, $desc);
        }
        return $usage;
    }
}

function cmp_searchoptions(SearchOption $o1, SearchOption $o2): int
{
    return strcmp($o1->sortarg, $o2->sortarg);
}

File: fsharp/FsSearch/FsSearch/SearchOptions.fs
namespace FsSearch

open System
open System.Text.Json
open System.Text.RegularExpressions
open System.Xml.Linq

module SearchOptions =   

    type SearchOption = {
        ShortArg : string;
        LongArg : string;
        Description : string
    }

    let AddExtensions (exts : string) (extList : string list) : string list =
        List.append extList (FileUtil.ExtensionsListFromString exts)

    let AddPattern (pattern : string) (patternList : Regex list) : Regex list =
        List.append patternList [Regex(pattern)]

    let FileTypesListFromString (fts : string) : FileType list =
        let nonWord = Regex(@"\W+")
        nonWord.Split(fts)
        |> Array.toList
        |> List.filter (fun (x : string) -> String.IsNullOrEmpty(x) = false)
        |> List.map (fun (x : string) -> FileTypes.FromName x)

    let AddFileTypes (fts : string) (ftList : FileType list) : FileType list =
        List.append ftList (FileTypesListFromString fts)

    let argActionMap : Map<string, string -> SearchSettings.t -> SearchSettings.t> =
        [
            ("encoding", (fun (s : string) (settings : SearchSettings.t) -> { settings with TextFileEncoding = s }));
            ("in-archiveext", (fun (s : string) (settings : SearchSettings.t) -> { settings with InArchiveExtensions = AddExtensions s settings.InArchiveExtensions }));
            ("in-archivefilepattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with InArchiveFilePatterns = AddPattern s settings.InArchiveFilePatterns }));
            ("in-dirpattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with InDirPatterns = AddPattern s settings.InDirPatterns }));
            ("in-ext", (fun (s : string) (settings : SearchSettings.t) -> { settings with InExtensions = AddExtensions s settings.InExtensions }));
            ("in-filepattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with InFilePatterns = AddPattern s settings.InFilePatterns }));
            ("in-filetype", (fun (s : string) (settings : SearchSettings.t) -> { settings with InFileTypes = AddFileTypes s settings.InFileTypes }));
            ("in-linesafterpattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with InLinesAfterPatterns = AddPattern s settings.InLinesAfterPatterns }));
            ("in-linesbeforepattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with InLinesBeforePatterns = AddPattern s settings.InLinesBeforePatterns }));
            ("linesafter", (fun (s : string) (settings : SearchSettings.t) -> { settings with LinesAfter = Int32.Parse(s) }));
            ("linesaftertopattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with LinesAfterToPatterns = AddPattern s settings.LinesAfterToPatterns }));
            ("linesafteruntilpattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with LinesAfterUntilPatterns = AddPattern s settings.LinesAfterUntilPatterns }));
            ("linesbefore", (fun (s : string) (settings : SearchSettings.t) -> { settings with LinesBefore = Int32.Parse(s) }));
            ("maxlinelength", (fun (s : string) (settings : SearchSettings.t) -> { settings with MaxLineLength = Int32.Parse(s) }));
            ("out-archiveext", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutArchiveExtensions = AddExtensions s settings.OutArchiveExtensions }));
            ("out-archivefilepattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutArchiveFilePatterns = AddPattern s settings.OutArchiveFilePatterns }));
            ("out-dirpattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutDirPatterns = AddPattern s settings.OutDirPatterns }));
            ("out-ext", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutExtensions = AddExtensions s settings.OutExtensions }));
            ("out-filepattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutFilePatterns = AddPattern s settings.OutFilePatterns }));
            ("out-filetype", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutFileTypes = AddFileTypes s settings.OutFileTypes }));
            ("out-linesafterpattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutLinesAfterPatterns = AddPattern s settings.OutLinesAfterPatterns }));
            ("out-linesbeforepattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with OutLinesBeforePatterns = AddPattern s settings.OutLinesBeforePatterns }));
            ("searchpattern", (fun (s : string) (settings : SearchSettings.t) -> { settings with SearchPatterns = AddPattern s settings.SearchPatterns }));
        ] |> Map.ofList

    let flagActionMap : Map<string, bool -> SearchSettings.t -> SearchSettings.t> =
        [
            ("allmatches", (fun (b : bool) (settings : SearchSettings.t) -> { settings with FirstMatch = not b }));
            ("archivesonly", (fun (b : bool) (settings : SearchSettings.t) -> SearchSettings.SetArchivesOnly b settings));
            ("colorize", (fun (b : bool) (settings : SearchSettings.t) -> { settings with Colorize = b }));
            ("debug", (fun (b : bool) (settings : SearchSettings.t) -> SearchSettings.SetDebug b settings));
            ("excludehidden", (fun (b : bool) (settings : SearchSettings.t) -> { settings with ExcludeHidden = b }));
            ("firstmatch", (fun (b : bool) (settings : SearchSettings.t) -> { settings with FirstMatch = b }));
            ("help", (fun (b : bool) (settings : SearchSettings.t) -> { settings with PrintUsage = b }));
            ("includehidden", (fun (b : bool) (settings : SearchSettings.t) -> { settings with ExcludeHidden = not b }));
            ("listdirs", (fun (b : bool) (settings : SearchSettings.t) -> { settings with ListDirs = b }));
            ("listfiles", (fun (b : bool) (settings : SearchSettings.t) -> { settings with ListFiles = b }));
            ("listlines", (fun (b : bool) (settings : SearchSettings.t) -> { settings with ListLines = b }));
            ("multilinesearch", (fun (b : bool) (settings : SearchSettings.t) -> { settings with MultiLineSearch = b }));
            ("nocolorize", (fun (b : bool) (settings : SearchSettings.t) -> { settings with Colorize = not b }));
            ("noprintmatches", (fun (b : bool) (settings : SearchSettings.t) -> { settings with PrintResults = not b }));
            ("norecursive", (fun (b : bool) (settings : SearchSettings.t) -> { settings with Recursive = not b }));
            ("nosearcharchives", (fun (b : bool) (settings : SearchSettings.t) -> { settings with SearchArchives = not b }));
            ("printmatches", (fun (b : bool) (settings : SearchSettings.t) -> { settings with PrintResults = b }));
            ("recursive", (fun (b : bool) (settings : SearchSettings.t) -> { settings with Recursive = b }));
            ("searcharchives", (fun (b : bool) (settings : SearchSettings.t) -> { settings with SearchArchives = b }));
            ("uniquelines", (fun (b : bool) (settings : SearchSettings.t) -> { settings with UniqueLines = b }));
            ("verbose", (fun (b : bool) (settings : SearchSettings.t) -> { settings with Verbose = b }));
            ("version", (fun (b : bool) (settings : SearchSettings.t) -> { settings with PrintVersion = b }));
        ] |> Map.ofList;

    type SearchOptionsDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<System.Collections.Generic.Dictionary<string,string>>>

    let OptionsFromJson (jsonString : string) : SearchOption list =
        let searchOptionsDict = JsonSerializer.Deserialize<SearchOptionsDictionary>(jsonString)
        let optionDicts = searchOptionsDict.["searchoptions"]
        [ for optionDict in optionDicts do
            let longArg = optionDict.["long"]
            let shortArg = if optionDict.ContainsKey("short") then optionDict.["short"] else ""
            let desc = optionDict.["desc"]
            yield { ShortArg=shortArg; LongArg=longArg; Description=desc } ]

    let OptionsFromXml (xmlString : string) : SearchOption list =
        let rec recOptionsFromXml (nodeList : XElement list) (options : SearchOption list) : SearchOption list =
            match nodeList with
            | [] -> options
            | n :: ns ->
                let short = [for a in n.Attributes(XName.Get("short")) do yield a.Value].Head
                let long = [for a in n.Attributes(XName.Get("long")) do yield a.Value].Head
                let desc = n.Value.Trim()
                recOptionsFromXml ns (List.append options [{ ShortArg=short; LongArg=long; Description=desc }])
        let optNodes = XDocument.Parse(xmlString).Descendants(XName.Get("searchoption"))
        recOptionsFromXml (List.ofSeq optNodes) []
        |> List.sortBy (fun o -> if (o.ShortArg <> "") then (o.ShortArg.ToLower() + "@" + o.LongArg) else o.LongArg)

//    let _searchOptionsResource = EmbeddedResource.GetResourceFileContents("FsSearch.Resources.searchoptions.xml");
    let _searchOptionsResource = EmbeddedResource.GetResourceFileContents("FsSearch.Resources.searchoptions.json");
//    let options = OptionsFromXml(_searchOptionsResource)
    let options = OptionsFromJson(_searchOptionsResource)

    let SettingsFromArgs (args : string[]) : SearchSettings.t * string =
        let optionNameMap =
            let shortargs = seq { for opt in options do if opt.ShortArg <> "" then yield (opt.ShortArg, opt.LongArg) }
            let longargs =  seq { for opt in options do yield (opt.LongArg, opt.LongArg) }
            Seq.append shortargs longargs
            |> Map.ofSeq

        let argRegex = Regex("^(?:-{1,2})(?<opt>.*)$")
        
        let (|IsOption|_|) (arg:string) =            
            let m = argRegex.Match(arg)
            if m.Success then Some(m.Groups.["opt"].Value) else None

        let rec recSettingsFromArgs (argList : string list) (settings : SearchSettings.t) : SearchSettings.t * string =
            match argList with
            | [] -> settings, ""
            | head :: tail ->
                match head with
                | IsOption opt ->
                    if optionNameMap.ContainsKey(opt) then
                        let long = optionNameMap.[opt]
                        if argActionMap.ContainsKey(long) then
                            match tail with
                            | [] ->
                                settings, sprintf "Missing value for option: %s" opt
                            | aHead :: aTail -> 
                                recSettingsFromArgs aTail (argActionMap.[long] aHead settings)
                        elif flagActionMap.ContainsKey(long) then
                            if long = "help" then
                                recSettingsFromArgs [] (flagActionMap.[long] true settings)
                            else
                                recSettingsFromArgs tail (flagActionMap.[long] true settings)
                        else
                            settings, sprintf "Invalid option: %s" opt
                    else
                        settings, sprintf "Invalid option: %s" opt
                | _ -> recSettingsFromArgs tail { settings with StartPath = head }
        recSettingsFromArgs (Array.toList args) { SearchSettings.DefaultSettings with PrintResults = true }

    let GetUsageString () : string =
        let optStringMap =
            [ for opt in options do
                let shortstring : string = 
                    if opt.ShortArg <> "" then "-" + opt.ShortArg + ","
                    else ""
                let longstring : string = "--" + opt.LongArg
                yield (opt.LongArg, shortstring + longstring) ]
            |> Map.ofList

        let optDescMap =
            [ for opt in options do
                yield (opt.LongArg, opt.Description) ]
            |> Map.ofList

        let optStrings : string list =
            optStringMap
            |> Map.toList
            |> List.map snd

        let longest = 
            optStrings
            |> Seq.map (fun o -> o.Length)
            |> Seq.max

        let format = " {0,-" + longest.ToString() + "}  {1}"

        let usageStrings =
            [for o in options do
                yield String.Format(format, optStringMap.[o.LongArg], optDescMap.[o.LongArg])]

        let usageString = 
            usageStrings
            |> List.append ["\nUsage:"; " fssearch [options] -s <searchpattern> <startpath>\n"; "Options:"] 
            |> String.concat "\n"
        usageString

    let Usage (exitCode : int) : unit =
        let usageString = GetUsageString()
        printfn "%s\n" usageString
        Environment.Exit(exitCode)
